src/odemis/driver/actuator.py

```diff
@@ -2179,3 +2182,516 @@ def terminate(self):
+
+    def _doReference(self, axes):
+        """"
+        Perform sample stage reference procedure
+        """
+        # Only reference if the lens stage is already referenced and in deactive position
+        if not model.hasVA(self._lensz, "referenced"):
+            raise ValueError("Lens has to be initially referenced.")
+        f = self._lensz.moveAbs({'z': self._lensz.getMetadata()[model.MD_FAV_POS_DEACTIVE]})
+        f.result()
+        # Reference the dependant stage then update position
+        f = self._stage.reference(axes)
+        f.result()
+        self._updatePosition()
+
+    @isasync
+    def reference(self, axes):
+        """Start the referencing of the given axes"""
+        self._checkReference(axes)
+        return self._executor.submit(self._doReference, axes)
+
+    def stop(self, axes=None):
+        """
+        Stops the motion of the underlying dependencies
+        axes (iterable or None): list of axes to stop, or None if all should be stopped
+        """
+        # Empty the queue (and already stop the stage if a future is running)
+        self._executor.cancel()
+        # Stop the underlying stage
+        logging.debug("Stopping all axes: %s", ", ".join(self.axes))
+        self._stage.stop(axes)
+        # Stop the focus as well
+        self._focus.stop()
+        try:
+            self._updatePosition()
+        except Exception:
+            logging.exception("Unexpected failure when updating position")
+
+    def terminate(self):
+        if self._executor:
+            self.stop()
+            self._executor.shutdown()
+            self._executor = None
+
+class LinkedHeightFocus(model.Actuator):
+    """
+    This wrapper is the child of LinkedHeightActuator, it controls the lens stage Z axis and provide the means to convert it to
+    focus value and vice versa. It also shares the same executor for the movement functions with the parent.
+    """
+    def __init__(self, name, role, parent, dependencies, rng, **kwargs):
+        """
+        :param name: (string)
+        :param role: (string)
+        :param parent (Component): the parent of this component, that will be in
+        .parent (the LinkedHeightActuator object)
+        :param dependencies (dict str -> actuator): axis name (in this
+        actuator) -> actuator to be used for this axis (Presumably the lens Z axis)
+        :param rng: the focus range
+        """
+        self.parent = parent
+        self.range = rng
+        axes_def = {
+            "z": model.Axis(unit="m", range=rng),
+        }
+
+        model.Actuator.__init__(self, name, role, parent=parent, axes=axes_def, **kwargs)
+        self._metadata[model.MD_HW_NAME] = "LinkedHeightFocus"
+        # Get the shared executor from the parent
+        self._executor = parent.executor
+
+        self.position = model.VigilantAttribute({},
+                                                unit="m", readonly=True)
+
+        self.updateMetadata({model.MD_POS_COR: -0.005})
```

Comment:
I added `MD_POS_COR` in here to use it `updatePosition`, but let me know if there's a better place for it. @pieleric